lib/scholar/manifold/tsne.ex

```diff
@@ -0,0 +1,302 @@
+      """
+    ],
+    num_iters: [
+      type: :pos_integer,
+      default: 500,
+      doc: ~S"""
+      Maximum number of iterations for the optimization.
+      """
+    ],
+    seed: [
+      type: :integer,
+      doc: """
+      Determines random number generation for centroid initialization.
+      If the seed is not provided, it is set to `System.system_time()`.
+      """
+    ],
+    init: [
+      type: {:in, [:random, :pca]},
+      default: :pca,
+      doc: ~S"""
+      Initialization of embedding.
+      Possible options are `:random` and `:pca`.
+      """
+    ],
+    metric: [
+      type: {:in, [:euclidean, :squared_euclidean, :cosine, :manhattan, :chebyshev]},
+      default: :squared_euclidean,
+      doc: ~S"""
+      Metric used to compute the distances.
+      """
+    ],
+    exaggeration: [
+      type: {:or, [:float, :pos_integer]},
+      default: 10.0,
+      doc: ~S"""
+      Controls how tight natural clusters in the original space are in the embedded space and
+      how much space will be between them. For larger values, the space between natural clusters will be larger in the embedded space.
+      """
+    ]
+  ]
+
+  @opts_schema NimbleOptions.new!(opts_schema)
+
+  @doc """
+  Fits tSNE for sample inputs `x`.
+
+  ## Options
+
+  #{NimbleOptions.docs(@opts_schema)}
+
+  ## Return Values
+
+  Returns the embedded data `y`.
+
+  ## Examples
+
+      iex> x = Nx.iota({4,5})
+      iex> seed = 42
+      iex> Scholar.Manifold.TSNE.fit(x, num_components: 2, seed: seed)
+      #Nx.Tensor<
+        f32[4][2]
+        [
+          [11.721418380737305, 0.0],
+          [74.18099975585938, 0.0],
+          [-24.942012786865234, 0.0],
+          [-60.960227966308594, 0.0]
+        ]
+      >
+  """
+  deftransform fit(x, opts \\ []) do
+    opts = NimbleOptions.validate!(opts, @opts_schema)
+    seed = Keyword.get_lazy(opts, :seed, &System.system_time/0)
+    fit_n(x, seed, opts)
+  end
+
+  defnp fit_n(x, seed, opts \\ []) do
+    {perplexity, learning_rate, num_iters, num_components, exaggeration, init, metric} =
+      {opts[:perplexity], opts[:learning_rate], opts[:num_iters], opts[:num_components],
+       opts[:exaggeration], opts[:init], opts[:metric]}
+
+    x = to_float(x)
+    {n, _dims} = Nx.shape(x)
+
+    y1 =
+      case init do
+        :random ->
+          key = Nx.Random.key(seed)
+
+          {y, _new_key} =
+            Nx.Random.normal(key, 0.0, 1.0e-4, shape: {n, num_components}, type: Nx.type(x))
+
+          y
+
+        :pca ->
+          Scholar.Decomposition.PCA.fit_transform(x, num_components: num_components)
+      end
+
+    p = p_joint(x, perplexity, metric)
+
+    {y, _, _, _, _} =
+      while {y1, y2 = y1, learning_rate, p, i = 2}, i < num_iters do
+        q = q_joint(y1, metric)
+        grad = gradient(p * exaggeration(i, exaggeration), q, y1, metric)
+        y_next = y1 - learning_rate * grad + momentum(i) * (y1 - y2)
+        {y_next, y1, learning_rate, p, i + 1}
+      end
```

Comment:
```suggestion
    {y, _, _, _} =
      while {y1, y2 = y1, learning_rate, p}, i <- 2..num_iters, unroll: opts[:learning_loop_unroll] do
        q = q_joint(y1, metric)
        grad = gradient(p * exaggeration(i, exaggeration), q, y1, metric)
        y_next = y1 - learning_rate * grad + momentum(i) * (y1 - y2)
        {y_next, y1, learning_rate, p}
      end
```

using generator we can allow for unrolling through the options

Reply:
not a blocker

Reply:
I wonder if we should allow users to unroll every loop in `Scholar`. Otherwise, why should we do this specifically here?

Reply:
I think that allowing for control over each loop is best

Reply:
Let's move forward with this PR without this, and maybe we can reconsider this library-wide afterwards